File: app/controllers/EBNFController.java
package controllers;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import javax.inject.Inject;

import controllers.core.ExerciseController;
import model.CorrectionException;
import model.Secured;
import model.ebnf.EBNFExercise;
import model.ebnf.EBNFExercise$;
import model.ebnf.Grammar;
import model.ebnf.Rule;
import model.ebnf.RuleParser$;
import model.ebnf.TerminalSymbol;
import model.ebnf.Variable;
import model.result.CompleteResult;
import model.result.EvaluationResult;
import model.user.User;
import play.data.DynamicForm;
import play.data.FormFactory;
import play.mvc.Result;
import play.mvc.Security.Authenticated;
import play.twirl.api.Html;

@Authenticated(Secured.class)
public class EBNFController extends ExerciseController<EBNFExercise, EvaluationResult> {

  @Inject
  public EBNFController(FormFactory factory) {
    super(factory, "ebnf", EBNFExercise$.MODULE$.finder());
  }

  public Result index() {
    return ok(views.html.ebnf.ebnfIndex.render(getUser(), finder.all()));
  }

  @Override
  protected CompleteResult<EvaluationResult> correct(DynamicForm form, EBNFExercise exercise, User user)
      throws CorrectionException {
    Map<String, String[]> data = request().body().asFormUrlEncoded();

    List<TerminalSymbol> terminals = Arrays.stream(data.get("terminals")[0].split(",")).map(TerminalSymbol::new)
        .collect(Collectors.toList());
    List<Variable> variables = Arrays.stream(data.get("variables")[0].split(",")).map(Variable::new)
        .collect(Collectors.toList());
    Variable start = new Variable(data.get("start")[0]);
    String[] rulesAsString = data.get("rule[]");
    System.out.println(rulesAsString);
    List<Rule> rules = Arrays.stream(rulesAsString).map(RuleParser$.MODULE$::parse).collect(Collectors.toList());

    Grammar grammar = new Grammar(terminals, variables, start, rules);

    System.out.println(grammar);

    return null;
  }

  @Override
  protected Html renderExercise(User user, EBNFExercise exercise) {
    return views.html.ebnf.ebnfExercise.render(user, exercise);
  }

  @Override
  protected Html renderResult(CompleteResult<EvaluationResult> correctionResult) {
    // TODO Auto-generated method stub
    return null;
  }

}

File: modules/ebnf/app/controllers/ebnf/EBNFController.scala
package controllers.ebnf

import javax.inject.Inject

import controllers.core.IdExController
import model.Secured
import model.ebnf.{EBNFExercise, EBNFResult}
import model.result.CompleteResult
import model.user.User
import play.data.{DynamicForm, FormFactory}
import play.mvc.Security.Authenticated
import play.twirl.api.Html

import scala.util.Try

@Authenticated(classOf[Secured])
class EBNFController @Inject()(f: FormFactory)
  extends IdExController[EBNFExercise, EBNFResult](f, EBNFExercise.finder, EBNFToolObject) {

  override def correctEx(form: DynamicForm, exercise: EBNFExercise, user: User): Try[CompleteResult[EBNFResult]] = ??? // FIXME


  //    val data = Controller.request.body.asFormUrlEncoded
  //
  //    val terminals = data.get("terminals")(0).split(" ").map(str => new TerminalSymbol(str.replace("'", ""))).toList
  //    val variables = data.get("variables")(0).split(",").map(new Variable(_)).toList
  //    val start = new Variable(data.get("start")(0))
  //    val replacementsAsStrs = data.get("rule[]")
  //
  //    System.out.println(replacementsAsStrs.toList.mkString("\n"))
  //
  //    val replacements = replacementsAsStrs.map(RuleParser.parseReplacement(_))
  //    val rules = replacements.map(new Rule(Variable("S"), _)).toList
  //
  //    val grammar = new Grammar(terminals, variables, start, rules)
  //
  //    System.out.println(grammar)
  //
  //    new EBNFCompleteResult(new EBNFResult(grammar))

  override def renderExercise(user: User, exercise: EBNFExercise): Html = views.html.ebnfExercise.render(user, exercise)

  override def renderResult(correctionResult: CompleteResult[EBNFResult]): Html = views.html.ebnfResult.render(correctionResult)

  override def renderExesListRest = new Html("")

}